### ISSUE TITLE
Rails 4.2.5 reassign value to serialized field after SerializationTypeMismatch

### ISSUE THREAD

**Issue by olavnyy** [rails/rails#23449]
In Rails 4.2.5 after assign incorrect value type to serialized attribute error ActiveRecord::SerializationTypeMismatch raised and no possibility to **assign a new value!**

```
class TestSerialize < ActiveRecord::Base
  serialize :parameters, Array
  serialize :first_name, Hash
  serialize :last_name, String
end


2.2.3 :002 > a = TestSerialize.new
 => #<TestSerialize id: nil, parameters: [], first_name: {}, last_name: "", created_at: nil, updated_at: nil> 
2.2.3 :003 > a.parameters
 => [] 
2.2.3 :004 > a.parameters = "test"
ActiveRecord::SerializationTypeMismatch: Attribute was supposed to be a Array, but was a String. -- "test"

2.2.3 :005 > a.parameters
ActiveRecord::SerializationTypeMismatch: Attribute was supposed to be a Array, but was a String. -- "test"

2.2.3 :006 > a
ActiveRecord::SerializationTypeMismatch: Attribute was supposed to be a Array, but was a String. -- "test"

2.2.3 :007 > a.parameters = []
ActiveRecord::SerializationTypeMismatch: Attribute was supposed to be a Array, but was a String. -- "test"

2.2.3 :008 > a.parameters
ActiveRecord::SerializationTypeMismatch: Attribute was supposed to be a Array, but was a String. -- "test"


ActiveRecord::SerializationTypeMismatch:
       Attribute was supposed to be a Array, but was a String. -- "test"
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/coders/yaml_column.rb:17:in `dump'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/type/serialized.rb:26:in `type_cast_for_database'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/type/mutable.rb:5:in `type_cast_from_user'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute.rb:110:in `type_cast'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute.rb:42:in `original_value'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute.rb:37:in `value'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute_set.rb:31:in `fetch_value'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute_methods/read.rb:93:in `_read_attribute'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute_methods.rb:379:in `clone_attribute_value'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute_methods/dirty.rb:123:in `old_attribute_value'
     # /usr/local/rvm/gems/ruby-2.2.3/gems/activerecord-4.2.5/lib/active_record/attribute_methods/dirty.rb:94:in `write_attribute'
```

Looks like previous value is not set to nil after error raised!

**Comment by sgrif** [comment]
This has already been fixed in Rails 5. Unfortunately, due to the nature of this problem, the fix was unable to be backported to 4.2.

**Comment by poombavai** [comment]
Getting the same error. Can someone explain how it is fixed?
